test/distributions/privatization/runtime/deferDeletion.chpl

```diff
@@ -0,0 +1,27 @@
+// Tests the Quiescent State-Based Reclamation algorithm in cases where deadlock occurs
+// when a thread that cannot pass a checkpoint is waiting on a write that is waiting on them.
+// The desired result is that the older privatized array should be deferred for deletion
+// to this spinning thread.
+
+use PrivatizationWrappers;
+
+config const numIters = 100000;
+
+var done : atomic bool;
+done.write(false);
+
+// Spin on atomic counter without calling checkpoint
+begin {
+	while !done.read() do ; 
+	writeln("Spinning Thread Finished!");
+}
+
+for i in 1 .. numIters do insertPrivatized(new object(), i);
+done.write(true);
+
+// no leaks
+for i in 1 .. numIters {
+  var c = getPrivatized(i);
+  delete c;
+  clearPrivatized(i);
+}
```

Comment:
needs newline at end of file. I'm happy to have this test.

Also, can you think of more tests? Something like this QSBR code should really get some more tests designed to exercise the tricky cases. Can you think of 3 more, that exercise it in slightly different ways? Can you write tests of the QSBR code directly, instead of only testing it through privatization?

Reply:
The thing is that most of the privatization tests will be sufficient enough testing for QSBR, there shouldn't be any similar 'GOTCHAS' outside of memory reclamation as writers are not deadlocked anymore, and on chpl_qsbr_exit we free all data so it won't really be noticed. I wonder if I can query about state of memory *before* exiting to ensure they are deleted.

Reply:
Even so, I think it'd be a good practice to have some QSBR-level tests. Especially for different use cases (e.g. deferring deletion of a list) that don't happen to be needed in privatization.

Regarding checking for memory before exiting, you can do that. I recommend you look at modules/internal/MemTracking.chpl and search for a test that 'use MemTracking' to find examples.

Reply:
I added a new test employing a lock-free queue (one in which normally required hazard pointers, seems to run extremely well and is much simpler now), but not yet on `MemTracking` yet, I'll look into adding the test soon. If the abstraction I have is correct, this could be revolutionary.

Reply:
With the new pass-the-buck way of handling deferred deletion, even RCU writes are inexpensive now.

Reply:
Btw don't merge yet, found a bug in my code